Feature/fullbatch linkgenerator

adocherty opened the pull request:
### Description
This implements link prediction for full-batch models (GCN, GAT, APPNP, PPNP).

There is currently only an example for GCN link prediction.  I could create similar demos for GAT, APPNP, PPNP by copying this demo (which is itself copied from the GraphSAGE link prediction demo) and changing a single line creating the model, but this seems not so useful. Alternatively I could modify this notebook to have multiple, user-selectable, models.

This also replaces all `.node_model` method usages in full-batch methods with `.build`.  There should be no calls to `.node_model` or `.link_model` in demos now.

#### Reviewer Checklist
- [ ] Code has no obvious errors
- [ ] GCN link prediction notebook works
- [ ] Other notebooks (node prediction) work
- [ ] Code is well commented
- [ ] Full batch generators and classes well documented for link prediction

adocherty commented:
This PR also implements #439

geoffj-d61 commented on tests/layer/test_link_inference.py:
```diff
@@ -25,6 +25,140 @@
+            )
+        )
+        assert li.eval() == pytest.approx(expected)
+        assert li.eval() == pytest.approx(0)
```
```suggestion
        assert expected == pytest.approx(0)
```
Perhaps put this assertion up above with the computation of `expected`?

geoffj-d61 commented on tests/layer/test_link_inference.py:
```diff
@@ -25,6 +25,140 @@
+        x_dst /= np.linalg.norm(x_dst)  # normalize x_dst
+
+        x_src = x_src.reshape(1, 1, self.d)
+        x_dst = x_dst.reshape(1, 1, self.d)
```
Code reuse - suggest a helper method (with or without the `reshape`, depending on usage).

adocherty replied:
Definitely crying out for some refactoring 👍

adocherty commented on stellargraph/layer/appnp.py:
```diff
@@ -373,6 +395,20 @@ def node_model(self):
 
         return x_inp, x_out
 
+    def link_model(self):
```
Good question, I'm moving towards wanting them to be deprecated but I don't think we are quite there yet.  Maybe the new team can look at that!

adocherty commented on stellargraph/layer/graph_attention.py:
```diff
@@ -760,20 +765,31 @@ def __init__(
+
+        else:
+            if not isinstance(generator, FullBatchGenerator):
+                print(generator)
```
Thanks, that's some uncaught debug code! I'll remove it.

adocherty commented on demos/link-prediction/gcn/cora-gcn-links-example.ipynb:
```diff
@@ -0,0 +1,756 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "prediction = LinkEmbedding(output_dim=1, activation=\"relu\", method=\"ip\")(x_out)"
```
Yes, this one is weird – sigmoid would be much nicer theoretically, but for some reason relu works better. We could investigate this more...

The dot product is, as you point out, restricted from being negative at least. Theoretically would be better to not be greater than one, but doing that doesn't change anything so I left it as relu.

youph replied:
I think this activation doesn't do anything, and might as well be "linear". The real work is happening in the "ip" bit - the inner product, aka cosine similarity between node embeddings (which are l2-normalised, so are vectors on a unit sphere).

geoffj-d61 replied:
Agreed, relu on a non-negative dot product is just linear.

adocherty commented on stellargraph/core/graph.py:
```diff
@@ -591,10 +603,10 @@ def type_for_node(self, node):
     @property
     def node_types(self):
```
The majority of this class does not have type annotation. Why is it important to add it for this function?

geoffj-d61 replied:
Seems that if we are going to make a point (in the docs) that the return type is a set, it might be useful to start `typing` that. Not important, just niggled at me for some reason.

adocherty replied:
I think we need to add it to the base `StellarGraph` class if we want that.  For some reason this method is not there.

adocherty commented on stellargraph/layer/link_inference.py:
```diff
@@ -71,17 +72,153 @@ def compute_output_shape(self, input_shape):
+            out = Dense(self.output_dim, activation=self.activation)(out)
+
+        # Reshape to output dimensions
+        # out = Reshape((self.output_dim,))(out)
```
The output dimension is only ignored if the method is 'ip'.

> What does it mean, for example, to have node embeddings with dimension 50 (say using mul) and an output dimension of 2?  

This means the embeddings will be multiplied to a link-embedding of 50 and an additional dense layer with 2 output units is added.  There is an argument to be made that the link embedding layer should only do the first thing, and then the user can add a manual dense layer on after the embedding layer. 

This is really where I was heading with splitting this layer out  so I've gone the step further and moved this dense layer to the `link_inference` function itself and the `LinkEmbedding` layer now has no requirements for `output_dim`.

youph replied:
So this is a breaking change, right? We need to make sure we say it in the CHANGELOG, and, ideally, make it fail gracefully (with an informative prompt)

adocherty replied:
This has not changed the functionality of `link_inference` only of `LinkEmbedding` which is introduced in this PR, so I think we are OK?

youph commented on stellargraph/layer/link_inference.py:
```diff
@@ -71,17 +72,153 @@ def compute_output_shape(self, input_shape):
+             * 'concat' -- concatenation,
+             * 'ip' or 'dot' -- inner product, :math:`ip(u,v) = sum_{i=1..d}{u_i*v_i}`,
+             * 'mul' or 'hadamard' -- element-wise multiplication, :math:`h(u,v)_i = u_i*v_i`,
+             * 'l1' -- l1 operator, :math:`l_1(u,v)_i = |u_i-v_i|`,
```
I was wondering why it's "L"!

adocherty replied:
Done.

adocherty commented:
@geoffj-d61 Could you take a look at this again and confirm I've addressed your comments?

geoffj-d61 commented:
@adocherty , there is still the misspelling of `adjacency` as `adjacnecy` in `stellargraph/mapper/full_batch_generators.py`.

Also, in `stellargraph/layer/link_inference.py` I'd prefer the docs to mention that `l1` refers to the `L1` operator (or should that really be `norm`?), and likewise for `l2` -> `L2`. It is a mathematically accepted convention to be capitalised (see, e.g., Wikipedia), despite being frequently ignored by software engineers. Minor point, but what the hey!

Apart from those points, seems good!

youph commented:
Also:
`stellargraph/layer/link_inference.py:239` has `activation="linear"`, whereas it should be `activation=output_act`.

youph commented:
The check in `stellargraph/layer/link_inference.py:154` fails, since `x.shape[self.axis]` yields `Dimension(2)` instead of 2!

youph commented:
- `stellargraph/layer/link_inference.py:200`: what is `axis`? Doc string?
- link_inference defines `edge_function` that calls `LinkEmbedding` passing `axis` argument to it with default value of 0, whereas in `LinkEmbedding` definition the default value of `axis` is -2. This is confusing, and actually leads to an error when the following cell is executed instead of cell 17 in `cora-gcn-links-example.ipynb` demo:
```prediction = link_classification(
        output_dim=1, output_act="relu", edge_embedding_method='ip'
    )(x_out)```

adocherty commented:
# PR 543
> This PR looks good, yet I've noticed the following issues:
> 
>     * `/demos/node-classification/ppnp/ppnp-cora-node-classification-example.ipynb`: the old version was used here, which had a bug in the scalable APPNP model - could you please use the updated version of this demo in `develop`?

Merged develop into this branch.

>     * `stellargraph.mapper.full_batch_generators.FullBatchGenerator` lacks a doc string. Also, one of its arguments is `teleport_probability` with default value of 0.1, which is only relevant when used with  PPNP/APPNP models - should it be a `kwarg` instead?

This was not implemented in this issue.  I think it's better to create a separate issue for this.

>     * `stellargraph/mapper/full_batch_generators.py`: why does it export `FullBatchGenerator`?

So that models can just check if a `FullBatchGenerator` is passed rather than checking `FullBatchNodeGenerator` or `FullBatchLinkGenerator`.  It makes the model type checking much neater. I think we should do this for other Generators too. In fact I think we can go towards making the *NodeGenerator and *LinkGenerator just a single class with a flag for if we want nodes or links.

>     * (more general than this PR) `stellargraph/layer/gcn.py:430`: this requires integer node indices for model output. What would happen if the indices are strings? Will it break? Do we need to specify `dtype` here? E.g., in `stellargraph/layer/graphsage.py:1016` the input `dtype` is not specified.

The tensorflow model selects outputs as array indices and these are required to be integers. These are mapped from the indices in the original graph (any hashable incl. strings) to integer indices by the full-batch generators, see mapper/full_batch_generators.py:174.

>     * `stellargraph/demos/node-classification/gat/gat-cora-node-classification-example.ipynb`, cell 34 should be changed to
>
> ```
> X = emb.squeeze()
> y = np.argmax(target_encoding.transform(node_data[["subject"]].to_dict('records')), axis=1)
> ```
> otherwise the embeddings are scrambled. Could be the case for other similar demos.

Good catch .. I reordered the nodes and forgot this had implications!  Strange the GCN demo does it this way and the ensemble and GAT demos do it this way.  At least now the FullBatchGenerator gets the node order from the features array through `StellarGraph.nodes_of_type` .. this is not ideal and we should look at ensuring a consistent node order through `StellarGraph.nodes` that matches the feature node order.

I'll make an issue about this.

>     * (minor) `stellargraph/demos/link-prediction/gcn/cora-gcn-links-example.ipynb`, after cell 11: the narrative suggests there should be a summary of G_train and G_test, but no summary is actually shown.

I’ve removed the comment.  I think we can gloss over that.

>     * `stellargraph/demos/link-prediction/gcn/cora-gcn-links-example.ipynb`:
>       * cell 17: results in `TypeError: ('Keyword argument not understood:', 'output_dim')`. I think the call should be to `link_classification` instead of `LinkEmbedding` here. Yet changing it to `link_classification(output_dim=1, output_act="relu", edge_embedding_method='ip')(x_out)` produces an error: `ValueError: Expecting a tensor of shape 2 along specified axis for link embedding`. In any case this call is incompatible with that in GraphSAGE demo `stellargraph/demos/link-prediction/graphsage/cora-links-example.ipynb`!

Fixed. The current `link_classification` function reshapes the output which fails in this case. Therefore, `LinkEmbedding` is required here. If we want to use the `link_classficiation` function here we would need to add a `reshape=False` option to the function. This is why I created the `LinkEmbedding` layer to only perform link embedding calculation and not do the other things in the functions which I think should be controlled by the user.

I can create an issue about this.

>       * cell 16: if changed to `gcn.node_model()`, only a warning is produced, yet the node model is built, which results in failure of the next cell with a confusing error `ValueError: Cannot infer num from shape (1, ?, 16)`. Should we throw an error instead of a warning if generator is incompatible with `node_model` or `link_model` method?

My opinion is not to throw an error. This would be if you know what you are doing and you do actually want to create a node model as well as a link model for some reason (for example node embeddings – although you would be better off creating a node generator as you'll need it too).

I'd also consider deprecating node_model/link_model altogether.

> Also: stellargraph/layer/link_inference.py:239 has activation=“linear”, whereas it should be activation=output_act.

This is not an error. The `LinkEmbedding` layer has linear activation and the main activation function is applied in the `Dense` or by itself in lines 244:246 to match the original implementation of the `link_classification` function.  This was required to address the previous review comments.

> The check in stellargraph/layer/link_inference.py:154 fails, since x.shape[self.axis] yields Dimension(2) instead of 2!

Converted to an integer before comparison.  However, for TensorFlow 1.14 these Dimension objects do match integers so I’m unsure why there is an error:
```
tf.__version__
> '1.14.0'

x = tf.zeros((10,5))
x.shape[1]
> Dimension(5)

x.shape[1] == 5
> True
```

> Also:
> stellargraph/layer/link_inference.py:200: what is axis? Doc string? Same applies to the LinkEmbedding layer.
> link_inference defines edge_function that calls LinkEmbedding passing axis argument to it with default value of 0, whereas in LinkEmbedding definition the default value of axis is -2. This is confusing, and actually leads to an error when the following cell is executed instead of cell 17 in cora-gcn-links-example.ipynb demo:>
> prediction = link_classification(
>         output_dim=1, output_act=“relu”, edge_embedding_method=‘ip’
>     )(x_out)
> Is there a way to avoid having the axis argument altogether?

This axis is only used if the node embeddings are contained in a single tensor (the case for full-batch methods)  and is the axis that we unstack the outputs.  I’ve removed the axis from link_inference and left it in `LinkEmbedding` and added a description to the docstring of all link functions to make clear that the second-to-last axis is expected as the node embedding index axis.

adocherty commented on stellargraph/layer/link_inference.py:
```diff
@@ -71,17 +72,153 @@ def compute_output_shape(self, input_shape):
+                    "Expecting a tensor of shape 2 along specified axis for link embedding"
+                )
+            x0, x1 = tf.unstack(x, axis=self.axis)
+
```
Added and raised a type error.

adocherty commented on stellargraph/layer/link_inference.py:
```diff
@@ -71,17 +72,153 @@ def compute_output_shape(self, input_shape):
+            "axis": self.axis,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
```
Good point, but I'm going to leave this one.

youph commented:
Additional comments:
- some new demos featuring full-batch models still have the old comments, e.g., `stellargraph/demos/link-prediction/gcn/cora-gcn-links-example.ipynb` has many references to GraphSAGE.

adocherty commented:
OK I think I know why this is happening.  I'll take a look!

adocherty commented:
>     * some new demos featuring full-batch models still have the old comments, e.g., `stellargraph/demos/link-prediction/gcn/cora-gcn-links-example.ipynb` has many references to GraphSAGE.

Good catch – fixed.
 
>     * the training history of GCN link predictor looks weird, especially the loss:
>       ![image](https://user-images.githubusercontent.com/9361173/71149258-23352600-2282-11ea-9ed5-b55bcb313f02.png)
>       Could there be the same issue that affects the full-batch node predictors, described above?

This has been fixed, and full-batch notebooks re-run, the outputs now look as expected.

>     * I'm still confused about the `axis` argument of the `LinkEmbedding` layer constructor. It is not used in `link_inference` anymore, the call of `LinkEmbedding` there uses the default value of `axis=-2`, so why do we need it as an argument at all? If it's needed for full use with batch methods, can we just define it under the hood from the generator type?

The axis argument is the axis that enumerates the nodes in the link and therefore the axis which is reduced when producing a link embedding.  It seems useful to have this able to be specified for the layer, but not in the functions that use the layer, e.g. `link_classification`.

adocherty commented:
I understand your concern, and I've made a new issue for this:
https://github.com/stellargraph/stellargraph/issues/711

I think we need to merge this PR and not do further development in it – I think that issue will be self-contained enough for a sg-library team member to take it on.

youph commented:
> I understand your concern, and I've made a new issue #711 for this. I think we need to merge this PR and not do further development in it – I think that issue will be self-contained enough for a sg-library team member to take it on.

Agreed. @huonw @kjun9 fyi ^

huonw commented:
> The docs build is still failing, please fix before merging.

Per the discussion on slack, it's not a problem with this PR and cannot be fixed here (https://github.com/readthedocs/readthedocs.org/issues/6566). We can and should merge ignoring it, unfortunately.
